### PR TITLE
🪟 🐛  Allow Free Connector Program syncs in workspace with no credits

### PR DESCRIPTION
**Pull request by joeykmh**
## What
Closes https://github.com/airbytehq/airbyte-internal-issues/issues/3829

Allows users to enable and manually sync connections irrespective of their workspaces `creditStatus` or `remainingCredits`. This avoids the issue where workspaces with 0 credits cannot start free syncs. This solution was discussed here: https://airbytehq-team.slack.com/archives/C049XDAS19C/p1675775965163619

## How
Instead of conditionally enabling pieces of the UI, this PR actually enables all manual sync buttons and "enable connection" switches. The backend will potentially reject these requests if the workspace credits do not suffice:

https://user-images.githubusercontent.com/7550957/217348791-fbf7bcb7-69e5-41bb-8e20-510a7fd9ed57.mov

I've also removed `FeatureItem.AllowSync`, since I removed the only usage of the feature.

### CODE REVIEW

File: airbyte-webapp/src/core/domain/catalog/traverseSchemaToField.test.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import { AirbyteJSONSchema } from "core/jsonSchema/types";
 
 import { traverseSchemaToField } from "./traverseSchemaToField";
 
-describe(`${traverseSchemaToField}`, () => {
+describe(`${traverseSchemaToField.name}`, () => {
```
Comment (josephkmh) [inline airbyte-webapp/src/core/domain/catalog/traverseSchemaToField.test.ts:5]: Unrelated to this PR, but prevents the stringified function from being logged during unit tests 😄

File: airbyte-webapp/src/packages/cloud/locales/en.json [hunk 2]
```diff
@@ -150,7 +150,7 @@
   "credits.emailVerification.resendConfirmation": "We send you a new verification link.",
   "credits.emailVerification.resend": "Send verification link again",
   "credits.lowBalance": "Your credit balance is low. You need to add more credits to prevent your syncs from being stopped.",
-  "credits.zeroBalance": "All your syncs have been stopped because your credit balance is 0. Add more credits to enable your syncs again.",
+  "credits.zeroBalance": "All your paid syncs have been stopped because your credit balance is 0. Add more credits to enable your paid syncs again.",
```
Comment (josephkmh) [inline airbyte-webapp/src/packages/cloud/locales/en.json:153]: @andyjih any thoughts on this wording? I want to make it clear that free connector program syncs will continue.

image

File: airbyte-webapp/src/hooks/services/useConnectionHook.tsx [hunk 3]
```diff
@@ -230,12 +230,23 @@ const useUpdateConnection = () => {
...
       updateConnection({ connectionId, status: enable ? ConnectionStatus.active : ConnectionStatus.inactive }),
     {
+      onError: (error: Error) => {
+        trackError(error);
+        registerNotification({
+          id: `tables.updateFailed.${error.message}`,
+          text: `${formatMessage({ id: "connection.updateFailed" })}: ${error.message}`,
+          type: ToastType.ERROR,
+        });
+      },
```
Comment (josephkmh) [inline airbyte-webapp/src/hooks/services/useConnectionHook.tsx:249]: Displays an error toast if enabling/disabling the connection failed (e.g. due to credit balance)

https://user-images.githubusercontent.com/7550957/217348029-9af939eb-f6ba-41c7-931d-b6d959b7fb34.mov

File: airbyte-webapp/src/packages/cloud/locales/en.json (hunk 2 above)
Comment (josephkmh) [reply airbyte-webapp/src/packages/cloud/locales/en.json:155]: Discussed here: https://airbytehq-team.slack.com/archives/C049XDAS19C/p1675775965163619

File: airbyte-webapp/src/components/EntityTable/components/StatusCell.tsx [hunk 4]
```diff
@@ -1,30 +1,30 @@
...
 
 export const StatusCell: React.FC<StatusCellProps> = ({
   enabled,
   isManual,
   id,
   isSyncing,
-  allowSync,
   schemaChange,
   hasBreakingChange,
+  connection,
```
Comment (tealjulia) [inline airbyte-webapp/src/components/EntityTable/components/StatusCell.tsx:27]: Now that this is receiving/passing along the entire connection object we can probably weed out some of these other props I would think.

Comment (tealjulia) [reply airbyte-webapp/src/components/EntityTable/components/StatusCell.tsx:27]: Also, it looks like these are a `ConnectionListItem` not `WebBackendConnectionRead`.  Maybe naming it `connectionListItem` would be more clear that we don't have an entire connection object here?

File: airbyte-webapp/src/components/EntityTable/components/StatusCellControl.tsx [hunk 5]
```diff
@@ -57,7 +56,7 @@ export const StatusCellControl: React.FC<StatusCellControlProps> = ({
         <Switch
           checked={enabled}
           onChange={onSwitchChange}
-          disabled={!allowSync || hasBreakingChange}
```
Comment (tealjulia) [inline airbyte-webapp/src/components/EntityTable/components/StatusCellControl.tsx:60]: This came from LaunchDarkly, right?  Was it currently targeting anyone?

File: airbyte-webapp/src/components/EntityTable/utils.tsx [hunk 6]
```diff
@@ -98,6 +98,7 @@ export const getConnectionTableData = (
     lastSyncStatus: getConnectionSyncStatus(connection.status, connection.latestSyncJobStatus),
     connectorIcon: type === "destination" ? connection.source.icon : connection.destination.icon,
     entityIcon: type === "destination" ? connection.destination.icon : connection.source.icon,
+    connection,
```
Comment (tealjulia) [inline airbyte-webapp/src/components/EntityTable/utils.tsx:101]: Is this pulling out and renaming a bunch of the properties from each ConnectionListItem, and then also returning a nested copy of the entire original ConnectionListItem object? 

Should we just pull out the additional property(/ies) that we want?

**Review by teallarson — COMMENTED** [review]
Testing locally, but had a few questions along the way.

File: airbyte-webapp/src/components/EntityTable/utils.tsx (hunk 6 above)
Comment (josephkmh) [reply airbyte-webapp/src/components/EntityTable/utils.tsx:101]: Yeah so the issue is that we actually need (almost) the full `WebBackendConnectionListItem` (or `WebBackendConnectionRead`) to call `syncConnection(connection)` in `StatusCellControl.tsx`. I would rather not pull out and drill through the additional 6 properties, instead just pass through `connection`.

Agreed that it's weird that we are pulling out and renaming a bunch of properties here. I was also surprised to come across that. Probably the better option would be to just pass through the `connection` object and do all this renaming closer to the consumer. I did not want to go into that refactoring in this PR though!

File: airbyte-webapp/src/components/EntityTable/components/StatusCell.tsx (hunk 4 above)
Comment (josephkmh) [reply airbyte-webapp/src/components/EntityTable/components/StatusCell.tsx:27]: > Now that this is receiving/passing along the entire connection object we can probably weed out some of these other props I would think.

True, I think we can reduce quite a few of them! I would prefer to address this in a follow-up PR if you don't mind. Filed an issue here: https://github.com/airbytehq/airbyte/issues/22600

> Also, it looks like these are a ConnectionListItem not WebBackendConnectionRead. Maybe naming it connectionListItem would be more clear that we don't have an entire connection object here?

Hm, I didn't think about that. I generally rely on typescript to tell me what the actual underlying type is. Agreed that people probably would assume it's a `WebBackendConnectionRead` by the variable name, but I would say that both `WebBackendConnectionListItem` and `WebBackendConnectionRead` describe a connection.

**Review by teallarson — COMMENTED** [review]
It looks like this is all the usages of `FeatureItem.AllowSync`.  Does that mean this should go away?: https://github.com/airbytehq/airbyte/blob/a6990bc1958c28fcc4718676f8fc5f9710cdb108/airbyte-webapp/src/packages/cloud/cloudRoutes.tsx#L48

**Comment by josephkmh** [comment]
> It looks like this is all the usages of FeatureItem.AllowSync. Does that mean this should go away?:

That's a really good point, and yes! I think there is no reason to keep it around. This PR is the appropriate place to get rid of it too - I'll add that.

File: airbyte-webapp/src/packages/cloud/cloudRoutes.tsx [hunk 7]
```diff
@@ -40,21 +37,7 @@ const CloudSettingsPage = React.lazy(() => import("./views/settings/CloudSetting
...
   const workspace = useCurrentWorkspace();
-  const cloudWorkspace = useGetCloudWorkspace(workspace.workspaceId);
-
-  useEffect(() => {
-    const outOfCredits =
-      cloudWorkspace.creditStatus === CreditStatus.NEGATIVE_BEYOND_GRACE_PERIOD ||
-      cloudWorkspace.creditStatus === CreditStatus.NEGATIVE_MAX_THRESHOLD;
-    // If the workspace is out of credits it doesn't allow creation of new connections
-    // or syncing existing connections.
-    setWorkspaceFeatures(outOfCredits ? ({ [FeatureItem.AllowSync]: false } as FeatureSet) : []);
```
Comment (ambirdsall) [inline airbyte-webapp/src/packages/cloud/cloudRoutes.tsx:53]: Enforcing the policy by dynamically toggling an app-wide feature setting feels like a really hacky use of the feature system anyway; not sad to see it go.

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/lib/model.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+import { ReleaseStage } from "core/request/AirbyteClient";
+
+export const freeReleaseStages: ReleaseStage[] = ["alpha", "beta"];
```
Comment (ambirdsall) [inline airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/lib/model.ts:3]: Extracting this from the free tag component file is 100% the right move, thanks. Extremely optional suggestion for the API:
```suggestion
export const isFreeReleaseStage = (stage: ReleaseStage): boolean => ["alpha", "beta"].includes(stage);
```

**Review by ambirdsall — COMMENTED** [review]
:ship:

**Review by ambirdsall — APPROVED** [review]

**Review by teallarson — APPROVED** [review]
tested locally, lgtm

**Pull request merged**